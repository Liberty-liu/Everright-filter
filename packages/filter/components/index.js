import _ from 'lodash-es'
const importModules = import.meta.glob('./*.vue', { eager: true })
const modules = {}
_.forIn(importModules, (func, path) => {
  modules[`${path.replace(/(.\/|\.vue)/g, '')}Component`] = func.default
})
export default modules
