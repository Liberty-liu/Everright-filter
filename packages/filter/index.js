import EverrightFilter from './index.vue'
import './style/index.js'
export {
  EverrightFilter
}
