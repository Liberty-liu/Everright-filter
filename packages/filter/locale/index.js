import _ from 'lodash-es'
// const importModules = import.meta.glob('./*.js', { import: 'setup' })
const importModules = import.meta.glob('./*.js', { import: 'default', eager: true })
const modules = {}
_.forIn(importModules, (func, path) => {
  const re = /[a-zA-Z0-9_-]*(?=\.js)/g
  modules[path.match(re)[0]] = func
})
export default modules
