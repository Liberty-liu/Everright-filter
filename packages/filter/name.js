export default {
  EVERRIGHTFILTER: 'EverrightFilter',
  FILTERITEM: 'FilterItem',
  FILTERRULE: 'FilterRule',
  PICKERCOMPONENT: 'PickerComponent',
  TRIGGERCOMPONENT: 'TriggerComponent',
  OPERATORCOMPONENT: 'OperatorComponent',
  NUMBERTYPE: 'NumberType',
  SELECTTYPE: 'SelectType',
  TEXTTYPE: 'TextType',
  TIMETYPE: 'TimeType',
  DATETYPE: 'DateType',
  REGIONTYPE: 'RegionType',
  CASCADERTYPE: 'CascaderType',
  DATECOMPONENT: 'DateComponent',
  CONSTRAINTCOMPONENT: 'ConstraintComponent',
  DAYHOURCOMPONENT: 'DayHourComponent',
  LOGICALOPERATORCOMPONENT: 'LogicalOperatorComponent'
}
