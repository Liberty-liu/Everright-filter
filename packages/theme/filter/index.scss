//@use '../base' as *;
$namespace: 'Everright-filter';
$component-width: (
  'trigger': 300px,
  'operator': 150px,
  'value': 300px
);
@import './Main';
@import './PickerComponent';
@import './FilterRule';
@import './FilterItem';
@import './LogicalOperatorComponent';
@import './DateType';
@import './DateComponent';
@import './DayHourComponent';
@import './TriggerComponent';
@import './OperatorComponent';
@import './ConstraintComponent';
@import './TextType';
@import './SelectType';
@import './NumberType';
@import './TimeType';
@import './RegionType';
@import './CascaderType';
@keyframes ERFILTER-ERROR-shakeError {
  0% {
    transform: translateX(0); }
  15% {
    transform: translateX(0.375rem); }
  30% {
    transform: translateX(-0.375rem); }
  45% {
    transform: translateX(0.375rem); }
  60% {
    transform: translateX(-0.375rem); }
  75% {
    transform: translateX(0.375rem); }
  90% {
    transform: translateX(-0.375rem); }
  100% {
    transform: translateX(0); } }
.ERFILTER-ERROR {
  animation-name: ERFILTER-ERROR-shakeError;
  animation-fill-mode: forwards;
  animation-duration: .6s;
  animation-timing-function: ease-in-out;
  &.el-input__wrapper,.el-input__wrapper {
    box-shadow: 0 0 0 1px #f56c6c inset !important;
  }
  &.el-button {
    border-color: #f56c6c;
  }
}
