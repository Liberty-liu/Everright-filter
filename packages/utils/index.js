import _ from 'lodash-es'
const importModules = import.meta.glob('./*.js', { eager: true })
const modules = {}
_.forIn(importModules, (func, path) => {
  Object.keys(func).forEach((key) => {
    modules[key] = func[key]
  })
})
export default {
  ...modules
}
