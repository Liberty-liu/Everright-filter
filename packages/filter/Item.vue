<script>
import { computed, ref, reactive, provide, inject, toRefs, onMounted, watch, onBeforeUnmount } from 'vue'
import NAME from '@ER/filter/name.js'
import _ from 'lodash-es'
import hooks from '@ER/hooks'
import Rule from './Rule.vue'
import LogicalOperatorComponent from './components/LogicalOperator.vue'
import utils from '@ER/utils'
export default {
  name: NAME.FILTERITEM
}
</script>
<script setup>
const ER = inject('Everright')
const {
  t,
  lang
} = hooks.useI18n()
const ERConstraint = inject('EverrightConstraint', '')
const emit = defineEmits(['del'])
const props = defineProps({
  id: {
    type: String
  },
  index: {
    type: Number
  },
  parent: {
    type: Array
  }
})
const ns = hooks.useNamespace(NAME.FILTERITEM)
const ruleRef = ref()
const optionContentRef = ref()
const state = reactive({
  logicalOperator: 'and',
  // rules: ER.props.type === 'quick-filter' ? [] : utils.generateItems(1)
  rules: /^(quick-search|quick-filter)$/.test(ER.props.type) ? [] : utils.generateItems(1)
})
const isShowOperator = computed(() => {
  return ER.props.type !== 'quick-filter' && state.rules.length > 1
})
const isInConstraint = !!ERConstraint
const {
  getData,
  setData
} = hooks.useCommon(NAME.FILTERITEM, {
  ...toRefs(state),
  ...toRefs(props),
  ruleRef
})
const addRule = () => {
  if (ER.props.canAddRule() !== false) {
    state.rules.push(...utils.generateItems(1))
  }
}
defineExpose({
  getData,
  setData,
  addRule () {
    state.rules.push(...utils.generateItems(1))
  },
  state
})
const operatorHeight = ref('auto')
const callback = (mutationsList) => {
  let curHeight = 0
  const nodes = optionContentRef.value.querySelectorAll(`.${ns.e('rule')}`)
  for (let i = 0; i < nodes.length - 1; i++) {
    const cur = nodes[i]
    curHeight += cur.offsetHeight
  }
  if (nodes.length > 1) {
    curHeight += 52
  }
  operatorHeight.value = curHeight === 0 ? 'auto' : curHeight
}
let observer = ''
if (!isInConstraint) {
  onMounted(() => {
    const config = { attributes: false, childList: true, subtree: true }
    observer = new MutationObserver(callback)
    observer.observe(optionContentRef.value, config)
  })
  onBeforeUnmount(() => {
    observer.disconnect()
    observer = null
  })
}
watch(state.rules, (newVal) => {
  if (!newVal.length) {
    emit('del')
  }
}, { deep: true })
const itemLabel = computed(() => {
  return lang.value === 'zh-cn' ? utils.digitalToChinese(props.parent.indexOf(props.id) + 1) : utils.digitalToEnglish(props.parent.indexOf(props.id) + 1)
})
</script>
<template>
  <div :class="[ns.b()]">
    <h2 v-if="!isInConstraint && ER.props.type === 'matrix'">{{ t(`er.${NAME.FILTERITEM}.itemLabel`) }} {{ itemLabel }}</h2>
    <div :class="['EverrightFilterOption']">
      <LogicalOperatorComponent
        v-if="!isInConstraint && isShowOperator"
        :height="operatorHeight"
        v-model="state.logicalOperator"/>
      <div
        ref="optionContentRef"
        :class="[ns.e('optionContent'), 'EverrightFilterOptionContent']">
        <TransitionGroup name="el-fade-in">
          <Rule
            ref="ruleRef"
            :class="!isInConstraint && ns.e('rule')"
            v-for="(item, i) in state.rules"
            :key="item"
            :id="item"
            :parentId="id"
            :parent="state.rules"
            :index="index + '-' + i"
          />
        </TransitionGroup>
        <el-button
          v-if="isInConstraint ? true : ER.isShowAdd.value"
          :class="[ns.e('add')]"
          @click="addRule"
          link
        >{{isInConstraint ? t(`er.${NAME.FILTERITEM}.addProp`) : t(`er.${NAME.FILTERITEM}.addCondition`)}}</el-button>
      </div>
    </div>
  </div>
</template>
