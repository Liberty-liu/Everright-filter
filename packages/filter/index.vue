<script>
import { defineProps, ref, reactive, computed, provide, getCurrentInstance, watch, nextTick, onMounted, isReactive, readonly, toRefs } from 'vue'
import LogicalOperatorComponent from './components/LogicalOperator.vue'
import _ from 'lodash-es'
import async from 'async'
import Item from './Item.vue'
import NAME from '@ER/filter/name.js'
import hooks from '@ER/hooks'
import utils from '@ER/utils'
import {
  useXhr
} from './xhr'
export default {
  name: NAME.EVERRIGHTFILTER
}
</script>
<script setup>
const emit = defineEmits(['listener'])
const props = defineProps({
  type: {
    type: String,
    default: 'linear',
    validator (value) {
      return /^(linear|matrix|quick-search|quick-filter)$/.test(value)
    }
  },
  httpPrams: {
    default () {}
  },
  isShowValidateState: {
    type: Boolean,
    default: true
  },
  lang: {
    type: String,
    default: 'zh-cn'
  },
  getOptions: {
    type: Function
  },
  getConditions: {
    type: Function
  },
  getProps: {
    type: Function
  },
  getPropValues: {
    type: Function
  },
  canAddRule: {
    type: Function,
    default: () => {}
  },
  ruleLimit: {
    type: Number,
    default: -1
  }
})
const {
  t,
  lang
} = hooks.useI18n(props)
const itemRef = ref()
const transitionName = ref('el-fade-in')
const state = reactive({
  loading: false,
  options: [],
  logicalOperator: 'and',
  store: {
    filters: utils.generateItems(1)
  },
  operators: {},
  Namespace: 'filter',
  children: [],
  remoteData: {},
  rulesState: {},
  showDels: {},
  pushRuleProperty: '',
  isShowAdd: /^(linear|matrix)$/.test(props.type)
})
const contentRef = ref()
const {
  getData,
  setData
} = hooks.useCommon(NAME.EVERRIGHTFILTER, {
  ...toRefs(state),
  transitionName,
  itemRef,
  props
})
const states = {}
const api = useXhr(props)
const readied = async.queue(function (task, callback) {
  states[String(task.name)] = {}
  task.fn.then(() => {
    callback()
    states[String(task.name)] = false
  })
})
readied.drain(function () {
  nextTick(() => {
    state.remoteData = {}
    transitionName.value = 'el-fade-in'
    state.loading = false
  })
})
let canFire = false
const fireEvent = (type, data) => {
  if (!canFire) {
    canFire = type === 'init'
  }
  if (canFire) {
    emit('listener', {
      type,
      data
    })
  }
}
const isShowAdd = computed(() => {
  // console.log(_.flatten(_.get(itemRef, 'value', []).map(e => e.state.rules)).length)
  return state.isShowAdd && props.ruleLimit === -1 ? true : _.flatten(_.get(itemRef, 'value', []).map(e => e.state.rules)).length < props.ruleLimit
})
provide('Everright', {
  state,
  api,
  readied,
  props,
  fireEvent,
  isShowAdd
})
const flatNodes = (nodes) => {
  return nodes.reduce((res, node) => {
    if (!node.children) {
      res.push(node)
    } else {
      res = res.concat(flatNodes(node.children))
    }
    return res
  }, [])
}
const initConfiguration = () => {
  nextTick(async () => {
    if (/^(linear|matrix)$/.test(props.type)) {
      state.loading = false
      fireEvent('init', _.cloneDeep(state.options))
      return false
    }
    if (props.type === 'quick-filter') {
      const pickers = state.children.filter(e => e.name === NAME.PICKERCOMPONENT)
      const nodes = _.cloneDeep(flatNodes(state.options))
      while (nodes.length) {
        itemRef.value[0].addRule()
        state.pushRuleProperty = nodes[0].value
        await nextTick()
        await nextTick()
        nodes.shift()
      }
      state.loading = false
      await nextTick()
      fireEvent('init', _.cloneDeep(state.options))
    }
    if (props.type === 'quick-search') {
      const pickers = state.children.filter(e => e.name === NAME.PICKERCOMPONENT)
      const nodes = _.cloneDeep(flatNodes(state.options))
      itemRef.value[0].addRule()
      state.loading = false
      await nextTick()
      fireEvent('init', _.cloneDeep(state.options))
    }
  })
}
const getRemoteData = async () => {
  state.loading = true
  try {
    const response = await api.getOptions()
    state.options = response.data.options
    state.operators = response.data.operators
    initConfiguration()
  } catch (e) {
  }
}
getRemoteData()
const pushData = (...arg) => {
  if (props.type === 'linear') {
    const isUniq = _.get(arg, '[1]', false)
    const pickers = state.children.filter(e => e.name === NAME.PICKERCOMPONENT)
    if (pickers.length === 1 && !pickers[0].state.property) {
      pickers[0].state.property = arg[0]
    } else {
      const findIndex = _.findIndex(pickers, {
        state: {
          property: arg[0]
        }
      })
      if (isUniq && findIndex !== -1) {
        return false
      }
      itemRef.value[0].addRule()
      state.pushRuleProperty = arg[0]
    }
  }
}
const typesRe = new RegExp(`^(${[NAME.NUMBERTYPE, NAME.TEXTTYPE, NAME.SELECTTYPE, NAME.TIMETYPE, NAME.DATETYPE, NAME.DATECOMPONENT, NAME.REGIONTYPE, NAME.CASCADERTYPE].join('|')})$`)
const clearData = (type) => {
  if (type === 'values') {
    const types = state.children.filter(e => typesRe.test(e.name))
    types.forEach(e => {
      e.clearData()
    })
  } else {
    state.store.filters = utils.generateItems(1)
    state.logicalOperator = 'and'
  }
}
defineExpose({
  getData,
  setData: (...arg) => {
    canFire = false
    state.loading = true
    setData(...arg)
    canFire = true
  },
  pushData,
  clearData
})
const ns = hooks.useNamespace('Main', state.Namespace)
const isShowOperator = computed(() => {
  return state.store.filters.length > 1
})
const addItem = () => {
  state.store.filters.push(...utils.generateItems(1))
}
const addGroupLabel = computed(() => {
  return `${t(`er.${NAME.EVERRIGHTFILTER}.addGroupLabel`)} ${lang.value === 'zh-cn' ? utils.digitalToChinese(state.store.filters.length + 1) : utils.digitalToEnglish(state.store.filters.length + 1)}`
})
</script>
<template>
  <div :class="[ns.b(), ns.e(props.type)]" v-loading="state.loading">
    <div v-if="type === 'matrix'" :class="['EverrightFilterOption']">
      <LogicalOperatorComponent
        v-if="isShowOperator"
        v-model="state.logicalOperator"
        :contentRef="contentRef"
      />
      <div v-if="state.options.length" :class="['EverrightFilterOptionContent']" ref="contentRef">
        <TransitionGroup :name="transitionName">
          <Item
            ref="itemRef"
            class="EverrightFilterItemSign"
            v-for="(item, index) in state.store.filters"
            :key="item"
            :id="item"
            @del="state.store.filters.splice(index, 1)"
            :parent="state.store.filters"
            :index="index"
          >
          </Item>
        </TransitionGroup>
        <el-button
          v-if="isShowAdd"
          :class="[ns.e('add')]"
          @click="addItem"
          link
        >{{ addGroupLabel }}</el-button>
      </div>
    </div>
    <div
      v-if="/^(linear|quick-search|quick-filter)$/.test(type)">
      <Item
        ref="itemRef"
        class="EverrightFilterItemSign"
        v-for="(item, index) in state.store.filters"
        :key="item"
        :id="item"
        @del="state.store.filters.splice(index, 1)"
        :parent="state.store.filters"
        :index="index"
      >
      </Item>
    </div>
  </div>
</template>
