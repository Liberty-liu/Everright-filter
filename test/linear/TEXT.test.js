import { describe, assert, expect, test, beforeEach, afterEach, beforeAll, vi } from 'vitest'
import { mount, flushPromises, enableAutoUnmount, config } from '@vue/test-utils'
import optionData from '@ER-server/routes/Filter/data/options.js'
import ElementPlus from 'element-plus'
import * as ElementPlusIconsVue from '@element-plus/icons-vue'
import { EverrightFilter } from '@ER/filter'
import NAME from '@ER/filter/name.js'
import utils from '@ER/utils'
import { nextTick } from 'vue'
import { getSelectOptions } from '../utils.js'
config.global.components.Delete = ElementPlusIconsVue.Delete
const _mount = (template, data, otherObj) => mount(
  {
    components: {
      EverrightFilter
    },
    template,
    data,
    ...otherObj
  },
  {
    attachTo: 'body',
    global: {
      plugins: [
        ElementPlus
      ]
    }
  }
)
const getOptions = async () => {
  return new Promise((resolve, reject) => {
    resolve({
      data: optionData
    })
  })
}
describe('renderType: TEXT', () => {
  let wrapper = {}
  const handleListener = vi.fn()
  beforeAll(() => {
    wrapper = _mount(`
      <everright-filter
        :getOptions="getOptions"
        ref="ERfilterRef"
      />
      `, () => ({
      handleListener,
      getOptions
    })
    )
  })
  afterEach(() => {
    wrapper.findComponent({ ref: 'ERfilterRef' }).vm.clearData()
  })
  test('operator: "style=noop" && value is not empty', async () => {
    await wrapper.find(utils.getTestId(`${NAME.PICKERCOMPONENT}-0-0`)).find('.Everright-filter-TriggerComponent').trigger('click')
    document.querySelector(utils.getTestId(`${NAME.TRIGGERCOMPONENT}-popperClass`, 'id')).querySelectorAll('.el-cascader-node')[0].click()
    await nextTick()
    document.querySelector(utils.getTestId(`${NAME.TRIGGERCOMPONENT}-popperClass`, 'id')).querySelectorAll('.el-cascader-menu')[1].querySelectorAll('.el-cascader-node')[5].click()
    await flushPromises()
    await wrapper.find(utils.getTestId(`${NAME.PICKERCOMPONENT}-0-0`)).find(utils.getTestId('operator', 'id')).trigger('click')
    expect(getSelectOptions(utils.getTestId('operator-popperClass', 'id')).map(e => utils.getAttrs(e).value)).toEqual([
      'equal',
      'one_of',
      'not_equal',
      'contains',
      'not_contain',
      'empty',
      'not_empty'
    ])
    let inputElm = wrapper.find(utils.getTestId(`${NAME.PICKERCOMPONENT}-0-0`)).find(utils.getTestId(`${NAME.TEXTTYPE}-input`, 'id'))
    const inputWrapper = inputElm.find('input')
    const nativeInput = inputWrapper.element
    expect(inputElm.exists()).toBe(true)
    expect(wrapper.findComponent({ ref: 'ERfilterRef' }).vm.getData()).toEqual({})
    await new Promise(resolve => setTimeout(resolve, 100))
    inputElm = wrapper.find(utils.getTestId(`${NAME.PICKERCOMPONENT}-0-0`)).find(utils.getTestId(`${NAME.TEXTTYPE}-input`, 'id'))
    expect(inputElm.classes()).toContain('ERFILTER-ERROR')
    nativeInput.value = '1'
    await inputWrapper.trigger('input')
    expect(wrapper.findComponent({ ref: 'ERfilterRef' }).vm.getData()).toMatchSnapshot()
  })
  test.only('operator: "style=one_of" && value is not empty', async () => {
    await wrapper.find(utils.getTestId(`${NAME.PICKERCOMPONENT}-0-0`)).find('.Everright-filter-TriggerComponent').trigger('click')
    document.querySelector(utils.getTestId(`${NAME.TRIGGERCOMPONENT}-popperClass`, 'id')).querySelectorAll('.el-cascader-node')[0].click()
    await nextTick()
    document.querySelector(utils.getTestId(`${NAME.TRIGGERCOMPONENT}-popperClass`, 'id')).querySelectorAll('.el-cascader-menu')[1].querySelectorAll('.el-cascader-node')[5].click()
    await flushPromises()
    await wrapper.find(utils.getTestId(`${NAME.PICKERCOMPONENT}-0-0`)).find(utils.getTestId('operator', 'id')).trigger('click')
    const selectOptions = getSelectOptions(utils.getTestId('operator-popperClass', 'id'))
    expect(selectOptions.map(e => utils.getAttrs(e).value)).toEqual([
      'equal',
      'one_of',
      'not_equal',
      'contains',
      'not_contain',
      'empty',
      'not_empty'
    ])
    selectOptions[1].click()
    await nextTick()
    let selectElm = wrapper.find(utils.getTestId(`${NAME.PICKERCOMPONENT}-0-0`)).find(utils.getTestId(`${NAME.TEXTTYPE}-select`, 'id'))
    expect(wrapper.findComponent({ ref: 'ERfilterRef' }).vm.getData()).toEqual({})
    await new Promise(resolve => setTimeout(resolve, 100))
    selectElm = wrapper.find(utils.getTestId(`${NAME.PICKERCOMPONENT}-0-0`)).find(utils.getTestId(`${NAME.TEXTTYPE}-select`, 'id'))
    expect(selectElm.classes()).toContain('ERFILTER-ERROR')
    const inputWrapper = selectElm.find('.el-select__input')
    const selectInputEl = inputWrapper.element
    selectInputEl.value = '1'
    vi.useFakeTimers()
    await inputWrapper.trigger('input')
    getSelectOptions(utils.getTestId(`${NAME.TEXTTYPE}-select-popperClass`, 'id'))[0].click()
    await nextTick()
    expect(wrapper.findComponent({ ref: 'ERfilterRef' }).vm.getData()).toMatchSnapshot()
  })
})
