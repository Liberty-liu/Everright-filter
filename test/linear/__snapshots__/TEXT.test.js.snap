// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renderType: TEXT > operator: "style=none" 1`] = `
{
  "filters": [
    {
      "conditions": [
        {
          "operator": "empty",
          "property": "text",
        },
      ],
      "logicalOperator": "and",
    },
  ],
  "logicalOperator": "and",
}
`;

exports[`renderType: TEXT > operator: "style=noop" && value is not empty 1`] = `
{
  "filters": [
    {
      "conditions": [
        {
          "operator": "equal",
          "property": "text",
          "value": "1",
        },
      ],
      "logicalOperator": "and",
    },
  ],
  "logicalOperator": "and",
}
`;

exports[`renderType: TEXT > operator: "style=one_of" && value is not empty 1`] = `
{
  "filters": [
    {
      "conditions": [
        {
          "operator": "one_of",
          "property": "text",
          "value": [
            "1",
          ],
        },
      ],
      "logicalOperator": "and",
    },
  ],
  "logicalOperator": "and",
}
`;
